docs: improve App Router internationalization guide

Vishwajeet-Kumar-Patel opened the pull request:
### What?

Enhances the App Router internationalization documentation with practical improvements for better developer experience.

### Why?

The current documentation, while functional, lacks:
- Installation instructions for required dependencies
- Best practices for SEO and locale persistence
- Common pitfall warnings
- TypeScript examples

These additions will help developers implement i18n more effectively and avoid common mistakes.

### How?

- Added installation commands for `@formatjs/intl-localematcher` and `negotiator`
- Added "Best Practices" section covering:
  - SEO optimization with `hreflang` tags
  - Locale persistence using cookies
  - TypeScript type definitions for dictionaries
  - Common pitfalls (async params, hydration issues, etc.)

### Testing

- Verified markdown syntax is correct
- Ensured code examples follow Next.js conventions
- Checked all code switchers (TypeScript/JavaScript) are consistent

Vishwajeet-Kumar-Patel committed:
docs: improve App Router internationalization guide

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel committed:
fix: apply prettier formatting

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +218,170 @@ export default async function RootLayout({ children, params }) {
+  params,
+}: LayoutProps<'/[lang]'>) {
+  const { lang } = await params
+
```
Let's align with the metadata API here

Vishwajeet-Kumar-Patel committed:
fix: align with metadata API type pattern

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel commented:
Hi @icyjoseph
Updated! Changed to use inline Props type definition with params: Promise<{ lang: string }> to align with the metadata API pattern. Thanks for the feedback!

icyJoseph commented:
That's not what I meant at all. I meant to use `export const metadata`

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
````diff
@@ -213,6 +218,172 @@ export default async function RootLayout({ children, params }) {
+For optimal SEO with internationalized content, add `hreflang` tags to indicate alternate language versions of your pages:
+
+```tsx filename="app/[lang]/layout.tsx" switcher
+type Props = {
````
Use `LayoutProps` here - but I am not sure about this PR to be honest, let's see

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +218,172 @@ export default async function RootLayout({ children, params }) {
+
+  return (
+    <html lang={lang}>
+      <head>
```
https://nextjs.org/docs/app/api-reference/functions/generate-metadata#alternates

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
````diff
@@ -213,6 +218,172 @@ export default async function RootLayout({ children, params }) {
+Consider storing the user's locale preference in a cookie to persist their choice across sessions:
+
+```tsx filename="proxy.ts" switcher
+import { NextResponse } from 'next/server'
````
This simplifies to

```tsx
export function proxy(request: NextRequest) {
  const savedLocale = request.cookies.get('NEXT_LOCALE')?.value
  const { pathname } = request.nextUrl

  if (savedLocale && !pathname.startsWith(`/${savedLocale}`)) {
    // Branch 1: REDIRECT
    return NextResponse.redirect(request.nextUrl)
  }

  // Branch 2: ALSO REDIRECT (always)
  return NextResponse.redirect(request.nextUrl)
}
```

🙅 or am I misreading?

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
````diff
@@ -213,6 +218,172 @@ export default async function RootLayout({ children, params }) {
+For better type safety with dictionaries, define types for your translations:
+
+```ts filename="app/[lang]/dictionaries.ts"
+import 'server-only'
````
Is this not replicating some other snippets we already have? and type casting 🤔 - can't TS infer the json data?

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +218,172 @@ export default async function RootLayout({ children, params }) {
+
+### Common Pitfalls
+
+- **Always await params**: In Next.js 15+, route parameters are promises and must be awaited. Forgetting to `await params` will cause runtime errors.
```
These feel a bit vague ~ 

The route params await is a non issue in TS, cuz these are typed manually or via the helper promises... I can see the reasoning behind though, but also, this guide is about i18n... we do mention the await params bit in other places, heavily.

I'd maybe keep the handle locale detection on the server idea. But I feel like it needs more development or reasoning about why. How to align the client and other bits too. Localization is not only about translations, but formatting etc.

Vishwajeet-Kumar-Patel committed:
docs: address PR review feedback on i18n guide

- Replace manual hreflang tags with Metadata API
- Fix locale persistence redirect logic bug
- Remove redundant TypeScript Support section
- Remove vague Common Pitfalls section

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel commented:
Hi @icyJoseph! Thank you for the thorough review and feedback. I've addressed all your comments:

SEO Optimization - Replaced the manual <head> tags with the Metadata API using generateMetadata and alternates.languages as you suggested
Type usage - Now using LayoutProps<'/[lang]'> throughout
Locale persistence - Fixed the broken redirect logic that was always redirecting. It now properly checks if the locale exists in the pathname first
Removed sections - Removed both the TypeScript Support section (which was duplicating existing content and using unnecessary type casting) and the Common Pitfalls section (which was too vague and not i18n-specific)
The guide now focuses purely on internationalization patterns and aligns better with Next.js documentation standards. Please let me know if there's anything else you'd like me to adjust!

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +218,138 @@ export default async function RootLayout({ children, params }) {
 
+## Best Practices
+
+### SEO Optimization
```
I think, this is not framed correctly, like, what's the SEO benefit of this. We just say that it is, and echo bits alraedy contained in other docs

Maybe, we can move this to the end of ## Routing Overview, as a new h3 heading, `### Alternate Language URLs`

````
### Alternate Language URLs

To help search engines discover your language-specific URLs, you can declare alternate versions using the Metadata API:

```tsx filename="app/[lang]/layout.tsx"
export const metadata = {
  alternates: {
    languages: {
      'en-US': 'https://example.com/en-US',
      'nl-NL': 'https://example.com/nl-NL',
      'x-default': 'https://example.com/en-US',
    },
  },
}
```

This generates `hreflang` tags that tell crawlers about available language variants. See [`alternates`](/docs/app/api-reference/functions/generate-metadata#alternates) for more options.
````

Something like that

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +218,138 @@ export default async function RootLayout({ children, params }) {
+
+let locales = ['en-US', 'nl-NL', 'nl']
+
+export function proxy(request: NextRequest) {
```
I do think here's a change to improve the first snippet on this guide w/ something like:

```tsx
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
import { match as matchLocale } from '@formatjs/intl-localematcher'
import Negotiator from 'negotiator'

const locales = ['en-US', 'nl-NL', 'nl']
const defaultLocale = 'en-US'

function getLocale(request: NextRequest): string {
  // To persist locale preference, check a cookie before negotiating:
  // const savedLocale = request.cookies.get('locale')?.value
  // if (savedLocale && locales.some((l) => l === savedLocale)) return savedLocale

  const negotiatorHeaders: Record<string, string> = {}
  request.headers.forEach((value, key) => (negotiatorHeaders[key] = value))

  const languages = new Negotiator({ headers: negotiatorHeaders }).languages(locales)
  return matchLocale(languages, locales, defaultLocale)
}
```

And then the second snippet we can add a mention to saving the locale to a cookie

```tsx
  const locale = getLocale(request)
  request.nextUrl.pathname = `/${locale}${pathname}`
  const response = NextResponse.redirect(request.nextUrl)
  // To persist locale preference, save to cookie:
  // response.cookies.set('locale', locale, { path: '/' })
  return response
```

Vishwajeet-Kumar-Patel replied:
Thanks for the detailed feedback! I've implemented both suggestions:

1. ✅ Updated the middleware snippets with proper imports (`NextResponse`, `NextRequest`, `matchLocale`, `Negotiator`), changed `let` to `const`, added `defaultLocale`, and included the complete `getLocale` function with commented cookie persistence logic

2. ✅ Moved the SEO section to the end of Routing Overview as "Alternate Language URLs" with clearer framing about helping search engines discover language-specific URLs

The changes are now pushed. Let me know if you'd like any adjustments!

Vishwajeet-Kumar-Patel committed:
docs: improve internationalization guide middleware examples and restructure SEO section

- Enhance proxy.js snippets with proper imports and type annotations
- Add defaultLocale constant and complete getLocale implementation
- Include comments for optional cookie-based locale persistence
- Move SEO content to 'Alternate Language URLs' section under Routing Overview
- Reframe hreflang documentation to focus on search engine discovery

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -18,29 +18,47 @@ It’s recommended to use the user’s language preferences in the browser to se
-let languages = new Negotiator({ headers }).languages()
-let locales = ['en-US', 'nl-NL', 'nl']
-let defaultLocale = 'en-US'
+const headers = { 'accept-language': 'en-US,en;q=0.5' }
```
My point was that this snippet is not helpful. Here we should've definite the getLocale function, with a comment that shows what headers look like w/ accept-language and then return the match result

icyJoseph commented on docs/01-app/02-guides/internationalization.mdx:
```diff
@@ -213,6 +252,73 @@ export default async function RootLayout({ children, params }) {
+    (locale) => pathname.startsWith(`/${locale}/`) || pathname === `/${locale}`
+  )
+  
+  if (pathnameHasLocale) return
```
I guess we can slash this whole Best Practices section then?

Vishwajeet-Kumar-Patel replied:
Good catch! I've updated both:

1. Replaced the match example with the full getLocale function definition, including a comment showing what headers look like with accept-language
2. Removed the Best Practices section entirely since cookie persistence is already covered in the main proxy example comments

Much cleaner now 👍

Vishwajeet-Kumar-Patel committed:
docs: refine i18n guide proxy examples and remove redundant Best Practices section

- Replace match example with getLocale function definition
- Add comment explaining headers structure with accept-language example
- Remove Best Practices section as cookie persistence is already documented in main proxy example comments

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel commented:
This PR only updates documentation. The skipped CI checks appear to be non-doc workflows that don’t run for docs-only changes. Let me know if you’d like me to adjust anything.

Vishwajeet-Kumar-Patel commented:
Hi team! 👋

This PR was approved by @kmk142789 2 days ago, but I notice it still needs an approval from a reviewer with write access to be merged. 

The PR only updates documentation - the skipped CI checks are non-doc workflows that don't run for docs-only changes.

Could someone with write access please review and merge when you get a chance? Thank you! 🙏

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel committed:
Update docs/01-app/02-guides/internationalization.mdx

Co-authored-by: vercel[bot] <35613825+vercel[bot]@users.noreply.github.com>

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n

Vishwajeet-Kumar-Patel committed:
Merge branch 'canary' into docs-app-router-i18n
